### PR TITLE
[SPARK-8975][Streaming] Adds a mechanism to send a new rate from the driver to the block generator

### PR DESCRIPTION
**Pull request by dragos**
First step for [SPARK-7398](https://issues.apache.org/jira/browse/SPARK-7398).

@tdas @huitseeker

### CODE REVIEW

File: streaming/src/test/scala/org/apache/spark/streaming/TestSuiteBase.scala [hunk 1]
```diff
@@ -537,4 +537,19 @@ trait TestSuiteBase extends SparkFunSuite with BeforeAndAfter with Logging {
...
   }
+
+  /**
+   * Wait until `cond` becomes true, or timeout ms have passed. This method checks the condition
+   * every 100ms, so it won't wait more than 100ms more than necessary.
+   *
+   * @param cond     A boolean that should become `true`
+   * @param timemout How many millis to wait before giving up
+   */
+  def waitUntil(cond: => Boolean, timeout: Int): Unit = {
```
Comment (dragos) [inline streaming/src/test/scala/org/apache/spark/streaming/TestSuiteBase.scala]: This method is general enough that probably deserves a better place. I couldn't find anything similar in spark-core Utils or streaming utilities (the closest one was in BatchCounter, but won't do the same thing, and relies on wait/notify). I'm open to suggestions where to add it.

File: streaming/src/main/scala/org/apache/spark/streaming/receiver/Receiver.scala [hunk 2]
```diff
@@ -271,7 +271,7 @@ abstract class Receiver[T](val storageLevel: StorageLevel) extends Serializable
   }
 
   /** Get the attached executor. */
-  private def executor = {
+  private[streaming] def executor = {
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/receiver/Receiver.scala]: Why this change? If you want to access private members, its best to use the `PrivateTester` See
https://github.com/apache/spark/blob/master/streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala#L866

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala [hunk 3]
```diff
@@ -180,6 +180,12 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
     logError(s"Deregistered receiver for stream $streamId: $messageWithError")
   }
 
+  /** Update a receiver's maximum rate from an estimator's update */
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala]: nit: There should be no references to estimator in this scala doc. Just "Update a receiver's max ingestion rate" is good enough.

File: streaming/src/test/scala/org/apache/spark/streaming/TestSuiteBase.scala (hunk 1 above)
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/TestSuiteBase.scala]: We have used `eventually` for doing this. See 
https://github.com/apache/spark/blob/master/streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala#L742

No need to add another method here.

File: streaming/src/test/scala/org/apache/spark/streaming/receiver/RateLimiterSuite.scala [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+package org.apache.spark.streaming.receiver
+
+import org.apache.spark.SparkConf
+import org.apache.spark.SparkFunSuite
+
+/** Testsuite for testing the network receiver behavior */
+class RateLimiterSuite extends SparkFunSuite {
+
+  test("rate limiter initializes even without a maxRate set") {
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/receiver/RateLimiterSuite.scala]: @dragos I had mentioned in the earlier PR that we need these test to actually send data through the rateLimiter to see whether the updated rate is taking effect. Otherwise this test to see test whether `a == 10` after setting `a = 10` is nice but not very useful.

File: streaming/src/main/scala/org/apache/spark/streaming/receiver/RateLimiter.scala [hunk 5]
```diff
@@ -34,12 +36,28 @@ import org.apache.spark.{Logging, SparkConf}
...
       rateLimiter.acquire()
     }
   }
+
+  private[receiver] def updateRate(newRate: Int): Unit =
+    if (newRate > 0) {
+      try {
+        if (maxRateLimit > 0) {
+          currentRateLimit.set(newRate.min(maxRateLimit))
+        }
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/receiver/RateLimiter.scala]: we use `} else {` format

File: streaming/src/main/scala/org/apache/spark/streaming/receiver/RateLimiter.scala [hunk 6]
```diff
@@ -34,12 +36,28 @@ import org.apache.spark.{Logging, SparkConf}
   */
 private[receiver] abstract class RateLimiter(conf: SparkConf) extends Logging {
 
-  private val desiredRate = conf.getInt("spark.streaming.receiver.maxRate", 0)
-  private lazy val rateLimiter = GuavaRateLimiter.create(desiredRate)
+  // treated as an upper limit
+  private val maxRateLimit = conf.getInt("spark.streaming.receiver.maxRate", 0)
+  private[receiver] var currentRateLimit = new AtomicInteger(maxRateLimit)
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/receiver/RateLimiter.scala]: What is this variable for? Seems like the only place its value is referenced outside this class is RateLimiterSuite. Rather than storing the current rate in the guava RateLimiter as well as this atomicInt, isnt it easier to just keep it one location, that is, the guava rate limiter? If you want to access the rate limit externally, then just have a def that returns, guava rateLimiter.getRate().

File: streaming/src/main/scala/org/apache/spark/streaming/receiver/Receiver.scala (hunk 2 above)
Comment (dragos) [inline streaming/src/main/scala/org/apache/spark/streaming/receiver/Receiver.scala]: I didn't know about `PrivateTester`, I'll look into it. Other Spark modules (such as core) are using these access modifiers, so I thought it's common practice.

File: streaming/src/test/scala/org/apache/spark/streaming/receiver/RateLimiterSuite.scala (hunk 4 above)
Comment (dragos) [inline streaming/src/test/scala/org/apache/spark/streaming/receiver/RateLimiterSuite.scala]: We could add an integration test. On the other hand, I personally don't see great use in testing Guava's code, I assume we take the same stance towards other libraries too. So, my assumption here is that *if* the update reaches the receiver, Guava's RateLimiter is working just fine.

File: streaming/src/test/scala/org/apache/spark/streaming/TestSuiteBase.scala (hunk 1 above)
Comment (dragos) [inline streaming/src/test/scala/org/apache/spark/streaming/TestSuiteBase.scala]: cool, thanks!

File: streaming/src/main/scala/org/apache/spark/streaming/receiver/Receiver.scala (hunk 2 above)
Comment (dragos) [inline streaming/src/main/scala/org/apache/spark/streaming/receiver/Receiver.scala]: `PrivateTester` won't work here. It assumes the private method is declared in the type of the receiver. It only works if the *runtime* type of the receiver is the type that declares the private method. Not the case here, where the method is declared in `Receiver`, but the invocation is on a concrete subclass (`DummyReceiver`). Relevant code [here](https://github.com/scalatest/scalatest/blob/master/scalatest/src/main/scala/org/scalatest/PrivateMethodTester.scala#L243).

I could probably code something along those lines, but I think it's better to use qualified private. Are there any downsides to relying on language support for visibility? I doubt any user-code would declare things in `o.a.s.streaming`.

**Comment by dragos** [comment]
See my last commit message about what changes I made for testing.

File: streaming/src/main/scala/org/apache/spark/streaming/receiver/RateLimiter.scala [hunk 7]
```diff
@@ -17,6 +17,8 @@
 
 package org.apache.spark.streaming.receiver
 
+import java.util.concurrent.atomic.AtomicInteger
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/receiver/RateLimiter.scala]: This is not needed any more. Make sure to remove the empty line below too.

File: streaming/src/test/scala/org/apache/spark/streaming/receiver/RateLimiterSuite.scala [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  def setupGenerator(blockInterval: Int): (BlockGenerator, FakeBlockGeneratorListener) = {
+    val blockGeneratorListener = new FakeBlockGeneratorListener
+    val conf = new SparkConf().set("spark.streaming.blockInterval", s"${blockInterval}ms")
+    val blockGenerator = new BlockGenerator(blockGeneratorListener, 1, conf)
+    (blockGenerator, blockGeneratorListener)
+  }
+
+  test("throttling block generator") {
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/receiver/RateLimiterSuite.scala]: This is a good attempt, but this is not the right attempt. This is a RateLimiterSuite, so ideally it should not be using something like a BlockGenerator to test the RateLimiter. So I would update the first three tests ("rate limiter **") to use a dummy rate limiter tester, that tests the upper limit that has been set, before and after any rate update has been made. The tester can be a simple integer increment. Note that only the upper limit needs to be tested, unlike the flaky block generator tests which test for both upper and lower limit.

Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/receiver/RateLimiterSuite.scala]: Ignore this. See comment above.

File: streaming/src/test/scala/org/apache/spark/streaming/receiver/RateLimiterSuite.scala (hunk 4 above)
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/receiver/RateLimiterSuite.scala]: I was not sure whether the Guava rate limiter behaves in the desired way when the rate is updated. There are several ways of implementing a updateable rate limiter, and I wanted to understand this's one behavior, and cover it in tests. However, I just digged deeper in Guava RateLimiter code, and realized that there are fundamentally bigger problems in using the Guava rate limiter, even for stable rate. It works effectively using token bucket philosophy, which allows a large burst of data to be allowed, if there has been no data for while. We DONT want that. This means that we need to implement our own rate limiter (which was the case, until someone replaced my implementation with Guava). Anyways, thats not something outside the scope of this PR. So I am okay not testing the rate limiting. 

So please revert ReceiverSuite, and remove the block manager unit tests in this testsuite. Basically, revert it to the previous iteration (sorry :) )
-

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala [hunk 9]
```diff
@@ -72,15 +77,50 @@ class ReceiverTrackerSuite extends TestSuiteBase {
...
+ */
+private object TestDummyReceiver extends DummyReceiver
+
 /**
  * Dummy receiver implementation
  */
 private class DummyReceiver(host: Option[String] = None)
   extends Receiver[Int](StorageLevel.MEMORY_ONLY) {
 
+  var started = false
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: This needs to be @volatile, otherwise waiting threads (one calling `eventually`) may never see the changed value ---> flakiness

File: streaming/src/main/scala/org/apache/spark/streaming/receiver/Receiver.scala (hunk 2 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/receiver/Receiver.scala]: Aah, I see. The reason I wanted to use PrivateTest is because marking private[streaming] make it invisible through Scala compilation, but its visibly publicly in the bytecode, and therefore through Java. And since Receivers can be written in Java, I dont want to expose that. PrivateTester wont work, but the underlying mechanism will working using reflection. 

```

scala> abstract class A { private val i = 1 }
defined class A

scala> class B extends A
defined class B

scala> B.getClass().getSuperclass()
<console>:8: error: not found: value B
              B.getClass().getSuperclass()
              ^

scala> classOf[B].getSuperclass()
res1: Class[_ >: B] = class A

scala> classOf[B].getSuperclass().getDeclaredFields()
res2: Array[java.lang.reflect.Field] = Array(private final int A.i)

scala> val f = classOf[B].getSuperclass().getDeclaredFields().head
f: java.lang.reflect.Field = private final int A.i

scala> f.setAccessible(true)

scala> f.get(new B)
res4: Object = 1


```

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala [hunk 10]
```diff
@@ -72,15 +77,50 @@ class ReceiverTrackerSuite extends TestSuiteBase {
...
+  def getCurrentRateLimit: Option[Long] = {
+    TestDummyReceiver.executor.getCurrentRateLimit
+  }
+}
+
+/**
+ * We need the receiver to be an object, otherwise serialization will create another one
+ * and we won't be able to read its rate limit.
+ */
+private object TestDummyReceiver extends DummyReceiver
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: I said this in another comment thread, but since that thread is hidden let me put it here as well. 

You could use reflection to get the private method "executor". Just make the DummyReceiver class have a method that gets the underlying supervisor using reflection. Then, instead of another class/object TestDummyReceiver, you can just use `object DummyReceiver extends DummyReceiver`

**Comment by tdas** [comment]
Things to change for the next iteration. 
- Revert block generate tests, and changes to ReceiverTracker (see inline threads for explanation)
- Revert Receiver class and modify DummyReceiver in ReceiverTrackerSuite to access the `executor` field using Java reflection (again, see other inline thread for discussion)

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala [hunk 11]
```diff
@@ -180,6 +180,12 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
     logError(s"Deregistered receiver for stream $streamId: $messageWithError")
   }
 
+  /** Update a receiver's maximum ingestion rate */
+  def sendRateUpdate(streamUID: Int, newRate: Long): Unit = {
+    for (info <- receiverInfo.get(streamUID); eP <- Option(info.endpoint))
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala]: nit: add `{` and `}` around the for-loop body

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala (hunk 10 above)
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: Why did you not remove TestDummyReceiver? You could only make a `object DummyReceiver extends DummyReceiver` and rename `TestReceiverInputDStream` to `DummyReceiverInputDStream`. No need to introduce another `Test***` (Test and Dummy is really redundant).

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala [hunk 12]
```diff
@@ -72,8 +78,62 @@ class ReceiverTrackerSuite extends TestSuiteBase {
     assert(locations(0).length === 1)
     assert(locations(3).length === 1)
   }
+
+  test("Receiver tracker - propagates rate limit") {
+    object streamingListener extends StreamingListener {
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: This should be capitalized ... say.. TestStreamingListener

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala [hunk 13]
```diff
@@ -72,8 +78,62 @@ class ReceiverTrackerSuite extends TestSuiteBase {
...
+      override def onReceiverStarted(receiverStarted: StreamingListenerReceiverStarted): Unit = {
+        started = true
+      }
+    }
+
+    ssc.addStreamingListener(streamingListener)
+    ssc.scheduler.listenerBus.start(ssc.sc)
+
+    val newRateLimit = 100L
+    val ids = new TestReceiverInputDStream(ssc)
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: ids --> dstream. I got confused with ids being IDs

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala [hunk 14]
```diff
@@ -72,8 +78,62 @@ class ReceiverTrackerSuite extends TestSuiteBase {
...
+    tracker.sendRateUpdate(ids.id, newRateLimit)
+    // this is an async message, we need to wait a bit for it to be processed
+    eventually(timeout(3 seconds)) {
+      assert(ids.getCurrentRateLimit.get === newRateLimit)
+    }
+  }
 }
 
+/** An input DStream with a hard-coded receiver that gives access to internals for testing. */
+private class TestReceiverInputDStream(@transient ssc_ : StreamingContext)
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: TestReceiverInputDStream --> DummyReceiverInputDStream

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala [hunk 15]
```diff
@@ -72,8 +78,62 @@ class ReceiverTrackerSuite extends TestSuiteBase {
...
+  private def invokeExecutorMethod: ReceiverSupervisor = {
+    val c = classOf[Receiver[_]]
+    val ex = c.getDeclaredMethod("executor")
+    ex.setAccessible(true)
+    ex.invoke(TestDummyReceiver).asInstanceOf[ReceiverSupervisor]
+  }
+}
+
+/**
+ * We need the receiver to be an object, otherwise serialization will create another one
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: nit: make it more scala docs style to begin with. "A receiver as an object, so that we can read its rate limit. Otherwise,.."

File: streaming/src/main/scala/org/apache/spark/streaming/receiver/RateLimiter.scala [hunk 16]
```diff
@@ -34,12 +34,26 @@ import org.apache.spark.{Logging, SparkConf}
...
+    rateLimiter.acquire()
   }
+
+  /**
+   * Return the current rate limit. If no limit has been set so far, it returns {{{Long.MaxValue}}}.
+   */
+  def getCurrentLimit: Long =
+    rateLimiter.getRate.toLong
+
+  private[receiver] def updateRate(newRate: Long): Unit =
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/receiver/RateLimiter.scala]: nit: Need a scala doc, for consistency

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala (hunk 10 above)
Comment (dragos) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: Is your issue only about naming? I personally find it weird to have the companion object of a class extend the class itself. It's true that I *could* do that, but I prefer not to. If you feel strongly about it I can do the renaming you wish, but I find it unnecessary.

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala (hunk 14 above)
Comment (dragos) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: I'll find a better name. IMO `Dummy` means nothing, so if names are important we could probably find a better one here. This isn't a dummy input dstream at all, it's a giving access to its rate limit.

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala (hunk 10 above)
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: Then let's call it SingletonReceiver. I find using test and dummy both in
the name very superfluous

On Wednesday, July 22, 2015, Iulian Dragos <notifications@github.com> wrote:

> In
> streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala
> <https://github.com/apache/spark/pull/7471#discussion_r35208599>:
>
> > +/** An input DStream with a hard-coded receiver that gives access to internals for testing. */
> > +private class TestReceiverInputDStream(@transient ssc_ : StreamingContext)
> > +  extends ReceiverInputDStream[Int](ssc_) {
> > +
> > +  override def getReceiver(): DummyReceiver = TestDummyReceiver
> > +
> > +  def getCurrentRateLimit: Option[Long] = {
> > +    TestDummyReceiver.executor.getCurrentRateLimit
> > +  }
> > +}
> > +
> > +/**
> > + * We need the receiver to be an object, otherwise serialization will create another one
> > + * and we won't be able to read its rate limit.
> > + */
> > +private object TestDummyReceiver extends DummyReceiver
>
> Is your issue only about naming? I personally find it weird to have the
> companion object of a class extend the class itself. It's true that I
> *could* do that, but I prefer not to. If you feel strongly about it I can
> do the renaming you wish, but I find it unnecessary.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/7471/files#r35208599>.
>


-- 
Sent from Gmail Mobile

Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: Then let's call it SingletonReceiver. I find using test and dummy both in
the name very superfl

On Wednesday, July 22, 2015, Iulian Dragos <notifications@github.com
<javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:

> In
> streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala
> <https://github.com/apache/spark/pull/7471#discussion_r35208599>:
>
> > +/** An input DStream with a hard-coded receiver that gives access to internals for testing. */
> > +private class TestReceiverInputDStream(@transient ssc_ : StreamingContext)
> > +  extends ReceiverInputDStream[Int](ssc_) {
> > +
> > +  override def getReceiver(): DummyReceiver = TestDummyReceiver
> > +
> > +  def getCurrentRateLimit: Option[Long] = {
> > +    TestDummyReceiver.executor.getCurrentRateLimit
> > +  }
> > +}
> > +
> > +/**
> > + * We need the receiver to be an object, otherwise serialization will create another one
> > + * and we won't be able to read its rate limit.
> > + */
> > +private object TestDummyReceiver extends DummyReceiver
>
> Is your issue only about naming? I personally find it weird to have the
> companion object of a class extend the class itself. It's true that I
> *could* do that, but I prefer not to. If you feel strongly about it I can
> do the renaming you wish, but I find it unnecessary.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/7471/files#r35208599>.
>


-- 
Sent from Gmail Mobile

**Comment by dragos** [comment]
Spurious failure:

========================================================================
Building Spark
========================================================================
[info] Building Spark (w/Hive 0.13.1) using SBT with these arguments:  -Pyarn -Phadoop-2.3 -Dhadoop.version=2.3.0 -Pkinesis-asl -Phive-thriftserver -Phive package assembly/assembly streaming-kafka-assembly/assembly streaming-flume-assembly/assembly
Using /usr/java/latest as default JAVA_HOME.
Note, this will be overridden by -java-home if it is set.
**Error: Invalid or corrupt jarfile build/sbt-launch-0.13.7.jar**

**Comment by dragos** [comment]
Same error :cry:

**Comment by tdas** [comment]
Jenkins, test this please

**Comment by tdas** [comment]
LGTM. Merging this
